scripts/annotate_sam_with_bed_features.py

```diff
@@ -0,0 +1,328 @@
+        ID: Attribute key to extract the feature identifier (default: "name").
+        extension: Nucleotides to adjust the feature coordinates (default: 0).
+
+    Returns:
+        A dictionary mapping alignment names to lists of feature tuples. If
+        the BED file is empty, returns None.
+    """
+    intersect_data = defaultdict(list)
+    Fields = namedtuple(
+        "Fields",
+        (
+            "feat_chr",
+            "source",
+            "feat_type",
+            "feat_start",
+            "feat_end",
+            "feat_score",
+            "strand",
+            "phase",
+            "feat_attributes",
+            "read_chr",
+            "read_start",
+            "read_end",
+            "read_name",
+            "read_score",
+            "read_strand",
+            "overlap_len",
+        ),
+    )
+
+    with open(intersect_file, "r", encoding="utf-8") as bedfile:
+        for line in bedfile:
+            fields = Fields(*line.strip().split("\t"))
+
+            feat_name = attributes_dictionary(fields.feat_attributes)[ID]
+            feat_start = int(fields.feat_start) + extension
+            feat_end = int(fields.feat_end) - extension
+
+            intersect_data[fields.read_name].append(
+                (feat_name, feat_start, feat_end)
+            )
+
+    if not intersect_data:
+        return None
+
+    return intersect_data
+
+
+def get_tags(
+    intersecting_feat: list, alignment: pysam.AlignedSegment, shift: int = 0
+) -> set:
+    """Construct a custom tag for an alignment based on intersecting features.
+
+    For each intersecting feature (given as a tuple of identifier, start, and
+    end), compute the 5'-shift and 3'-shift values relative to the alignment:
+        - 5'-shift = (alignment start - feature start + 1)
+        - 3'-shift = (alignment end - feature end)
+
+    The tag for a feature is generated in the format:
+
+        FEATURE_ID|5'-shift|3'-shift|CIGAR|MD|READ_SEQ
+
+    Only features where both shift values are within +/- `shift` are included.
+    If multiple features are valid, the resulting tags are collected in a set.
+
+    Args:
+        intersecting_feat: list of tuples (feature identifier, start and end).
+        alignment: A pysam.AlignedSegment object
+        shift: Maximum allowed absolute shift for both ends (default: 0).
+
+    Returns:
+        A set of tag strings constructed from valid intersecting features.
+    """
+    cigar = alignment.cigarstring
+    seq = alignment.query_sequence
+    md = alignment.get_tag("MD")
+
+    tags = []
+
+    for feat_name, feat_start, feat_end in intersecting_feat:
+        shift_5 = alignment.reference_start - feat_start + 1
+        shift_3 = alignment.reference_start + alignment.query_length - feat_end
+        if -shift <= shift_5 <= shift and -shift <= shift_3 <= shift:
+            tags.append(f"{feat_name}|{shift_5}|{shift_3}|{cigar}|{md}|{seq}")
+
+    return set(tags)
```

Comment:
Do you really need a set here? Doesn't the code already ensure that the values in tags are unique?